[WIP] feat: Icon component default styles and new class

EvilAlexei opened the pull request:
## Related Issue
Closes SAP/fundamental-styles#1710

## Description

BREAKING CHANGE:
Icon class changed from `.sap-icon` to `.fd-icon`

Changes:
- Renamed main class of Icon component from `sap-icon` to `fd-icon`
- Renamed all cases of `sap-icon` class to `fd-icon`
- Default styles according to a https://app.gitbook.com/@fundamental-styles/s/docs/icon#styling.
  - font-size: 1rem;
  - color: var(--sapContent_NonInteractiveIconColor)
  - background-color: var(--sapBackgroundColor);
- Added `Styles` section to stories
#### Note:
by documentation background-color should have different value `--sapContent_IconColor`, but this breaks design,
after discussion with @JKMarkowski it was decided to use `--sapBackgroundColor` as a temporary solution

## Screenshots
### Before:
![prev-sizes](https://user-images.githubusercontent.com/3996023/95858569-f6e51b80-0d65-11eb-9844-ea1f9301ea8e.png)
![prev-icons](https://user-images.githubusercontent.com/3996023/95858571-f8164880-0d65-11eb-9819-7f881f3ebadc.png)

### After:
![new-sizes](https://user-images.githubusercontent.com/3996023/95858574-f8164880-0d65-11eb-88da-5b63d9bd37ca.png)
![new-styles](https://user-images.githubusercontent.com/3996023/95858577-f8aedf00-0d65-11eb-87b0-93738e1c4efd.png)
![new-icons](https://user-images.githubusercontent.com/3996023/95858579-f8aedf00-0d65-11eb-97ac-decef81e1be1.png)

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option
- [x] Verified all styles in IE11
- [x] Updated tests
4. Documentation
- [x] Storybook documentation has been created/updated

BREAKING CHANGE:
Icon class changed from `.sap-icon` to `.fd-icon`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SAP/fundamental-styles?pullRequest=1778) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SAP/fundamental-styles?pullRequest=1778) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SAP/fundamental-styles?pullRequest=1778) it.</sub>

jbadan commented on src/fd-icon.scss:
```diff
@@ -0,0 +1,22 @@
+
+$block: #{$fd-namespace}-icon;
+
+[class*="#{$block}"] {
```
We're trying to move away from `[class*="#{$block}"] `, can this just reference the block like on line 16?

EvilAlexei replied:
Changed

EvilAlexei replied:
Reverted. 
We can't remove this part, because most of the icons doesn't have common unique class,
all of them looks like `.sap-icon--washing-machine` or `.sap-icon--wrench`.

It could be a good idea to add this class in future, for example as in the font-awesome:
```
<i class="fas fa-arrow-right"></i>
<i class="fas fa-arrow-left"></i>
```
I think this update shouldn't be a part of this PR.

jbadan replied:
That sounds reasonable! Can we make a followup issue for that? All places using an icon should probably be something like: `<i class='fd-icon fd-icon--iconName" />`

EvilAlexei replied:
@jbadan could you review please this issue :)
https://github.com/SAP/fundamental-styles/issues/1794

jbadan reviewed (changes requested):
Does this need to be a new icon component? Could we not just replace https://github.com/SAP/fundamental-styles/blob/master/src/icon.scss + documentation so that we don't have 2 icons at once?

Other than those small nits this looks good to me!

EvilAlexei commented:
@jbadan at this moment I think we can't just replace it because class `.sap-icon` already in use in many places,
for example, the button component uses it as well as the icon component himself.
So to replace the class name there should be added changes in related styles and related components in fundamentals-ngx.
In my opinion, these changes are breaking changes and should be done in this way, like creating a similar component, and then replacing it.

EvilAlexei commented:
@jbadan I always for the easier way :laughing: In my opinion, this way is the easiest, but I'm open to an alternative opinion too as well.

I see it this way: 
All libraries such as fundamentals-ngx are using this old class `sap-icon`.
If we change class in styles library and we do not change it in related libraries all cases with icons will be broken.
So we need to create separate PRs for each related library, and it can't be done in one PR, 
and they should be merged at the same time as PR for the styles library, otherwise all cases with the sap-icon will be broken.
I'm not really seeing it's possible to merge all PR's at the same time.

If I'm missing anything, please correct me.

jbadan commented:
@EvilAlexei The implementation libraries should be keeping their dependency on fundamental-styles up to date and keeping up with breaking changes.

In fundamental-react, we bump the fundamental-styles version every 2 weeks or so (and I hope ngx is doing something more or less similar). Since this change would be marked as a breaking change, the libraries shouldn't get this change unless they explicitly bump their version to include this breaking change - which will be called out in the release notes. The pr to bump fundamental-styles in fundamental-react or fundamental-ngx would also include all the changes that need to be made in the implementation library to bring this change in. In fundamental-react's case, this is a pretty simple change because we're using a component for icons and very rarely referring directly to `sap-icon--*`. It's going to be a find and replace for us (and hopefully for ngx unless something else is going on).

By making a breaking change in fundamental styles, you're not on the hook for making the changes in the implementation libraries - while it's nice when someone does do them all, it's the maintainers responsibility there (unless you are also a maintainer of ngx).

TLDR: We can do this asynchronously, there's no need for it to be done at the exact same time in all the libraries. Granted, we will be blocked from bringing in new features or fixes that are merged after this pr, but that's the nature of it and why we should be keeping the fundamental-styles dependency up to date as much as possible.

EvilAlexei commented:
@jbadan updated PR according to the last comments, removed new component `fd-icon` and transferred all new changes to the `icon` component.

Also reverted changes you mentioned earlier: `[class*="#{$block}"]`, answered in the original comment why I reverted this.

droshev reviewed (changes requested):
@EvilAlexei thank you for your hard work! we need to discuss how we want to continue with the icons before moving forward with this PR

jbadan commented:
> @jbadan updated PR according to the last comments, removed new component `fd-icon` and transferred all new changes to the `icon` component.
> 
Thank you for doing that - I know it was a lot of extra work :) Other than needing to have a more global group conversation about the direction of icons, I think this work looks good.

InnaAtanasova commented:
Let's park this for now and discuss it later.
